### ISSUE TITLE
Copy-paste tags

### ISSUE THREAD

**Issue by yvecai** [openstreetmap/iD#839]
A control should allow copying tag from one feature to another. Controls in the tag editor could do the trick.

**Comment by pauliuszaleckas** [comment]
+1 me too. I miss 'R' from Potlatch very much.

**Comment by brycenesbitt** [comment]
"Turn into area" seems more iD-like, and meets mostly the same use case.

**Comment by nighto** [comment]
@brycenesbitt no, it is a similar but not the same use case.

Let's say I want to add dozens of the same feature - bus stops in a street, for instance. They all have benches (so bench=yes) and are covered (so covered=yes). On Potlatch, JOSM, Vespucci etc. I could create the first node with all features I want, then create the following nodes and use the function to copy attributes (on Potlatch, press ```R```). On iD, I have to manually copy all attributes to the different nodes.

**Comment by HolgerJeromin** [comment]
vespucci has a nice GUI for selective copy and paste of tags in its newest version. I already miss this in JOSM :-)
A checkmark before each tag and one on top for all.
http://imgsdown.1mobile.com/group3/M00/20/93/S34UR1Xp-aGAQbauAAGDUSXQ64Q549.png
having some tags selected i could copy or cut the tags with buttons and can paste them with a button into another object (when there is no checkmark).

**Comment by nvkelso** [comment]
+1, especially helpful to correct tags.

**Comment by slhh** [comment]
Copy tags is a quite useful operation, but there are some risks associated with it:
* The souce object might have additional tags, which should not be copied.
* The selection might contain objects of wrong type (like vertices) accidentially. This is especially an issue in case of area based selections like lasso selection.
* Conflicting values might be accidentially overwritten.

I have proposed #3028 as a substitue in order to reduce these risks by giving the user more contol making the user more aware of the performed operations.

**Comment by jarlarntzen** [comment]
@slhh #3028 sounds good but I would also argue that a good start would be to replicate the existing "R"-shortcut key from Potlatch which simply repeats the tags of the previously selected object of the same type.

Point tags are repeated on points.
Line tags are repeated on lines.
Area tags are repeated on areas.

Conflicting tags are replaced, all others are kept. Nice for setting building=house where there is already a street address on each building.

Fast and nimble in use. Simple(r) to program.

**Comment by slhh** [comment]
@jarlarntzen 
> "R"-shortcut key from Potlatch which simply repeats the tags of the previously selected object of the same type.

This is a copy and paste tags function with an implicit copy part by selection. This doesn't reduce any risk of copy and paste tags but adds some risk of using the wrong object as source just due to a wrong click. Using the wrong object as source increases the risk of spreading or overwriting a tag unintentionally.
If iD users were working with all fields section closed and all tags section open, they would probably detect that something went wrong, provided that the operation is affecting a single object only.
Unfourtunately we can't assume this, because iD is mainly designed to be used differently.
If the iD user are working with all fields section open, unintended changes will be potentially hidden in the all tags section out of sight of the user.
For this reason copy and paste tags is more critical in iD than in other editors.

What seems to be safe even in iD is a modified and restricted version of the Potlatch "R"-key:
- Source object is the last object which the user has been tagged from scratch (within the current edit session); therefore, the user should know about used tags (there can't be hidden tags the user is unaware of). 
- Destination object is totally untagged or at least doesn't have any conflicting tags.
- Multiple destination objects must be limited to same basic type (node/way/relation). 
 
This would be a nice tool for some limited usecases, but we should have #3028 in adition to cover the remaining usecases.

**Comment by jarlarntzen** [comment]
@slhh I think this may be over-complicating a little bit. 

I agree that it could be beneficial with both an explicit "copy tags" action as well as a "paste tags" action that also shows a table-like summary of what tags are pasted or overwritten. As long as tags are pasted on one target at a time, (which is what Potlatch - `R` does) this will work fine.

If anything is messed up we always have `Ctrl+Z`.

Please try out `R` and `Shift + R` in Potlatch 2 to get a better sense of what it does. This is not the same as selecting and applying tags on multiple targets like in JOSM. 

Based on my experience with mapping tools over the last 5 years, I can at least testify that any work involving tracing and tagging houses, roads, pedestrian crossings, roads, paths, residential roads (anything repeated, really) is significantly sped up when one has the option of repeating / copying / pasting tags. This also includes correcting small mistakes, i.e. correcting the value of a tag.

Try houses or similar.

**Comment by slhh** [comment]
> If anything is messed up we always have Ctrl+Z.

This is sufficient provided that the user is aware that something has been messed up.
Otherwise the existence of the undo function is useless.

Let's assume we have a major road, which has been mapped quite well including turn:lanes, but the road name is missing on some of its OSM ways.
A user might want to correct the missing names using such an "R"-operation.
The result would potentially be spreading up the turn:lanes tag to other ways where they are not applicable.
The turn:lanes tags are not visible in the all fields section and are not rendered. Therefore, an unexperienced user would likely not be aware that something has been messed up. How could he know that he has to check all tags section carefully too? It is iDs concept of not requiring the user to read a thick manual before he can start to edit.

Btw, I had started with using P2 until I switched to JOSM after some month. If I had have to use iD instead of P2, I would have switched to JOSM much earlier. This is mainly due to missing "R"-operation or any other copy tags operation in iD. I know the P2 "R"-operation to be very useful, but to have some associated risks too. In P" the risk is manageable when using the advanced display, which is similar to using iD with collapsed all fields section and visible all tags section.

**Comment by yaugenka** [comment]
+1 for this feature. If you are mapping repeated objects you currently just have to switch to Potlatch or JOSM if you don't wan't to spend hours on typing the same tags over and over again.

**Comment by brycenesbitt** [comment]
Visibility is the issue.  iD for a variety of good reasons hides many nuts and bolts of operations.  Here a keyboard shortcut takes what may be a barely visible change.  While that change can be undone, that's assuming it's noticed in the first place.

**Comment by yaugenka** [comment]
How about adding copy and paste action icons right above the All Tags section? They way you will definitely see what you are copying and pasting.

**Comment by pnorman** [comment]
> How about adding copy and paste action icons right above the All Tags section? They way you will definitely see what you are copying and pasting.

Buttons also makes it easy to make the shortcuts discoverable since you can mouse-over them and have them say `Shift-^ V` or whatever you want.

**Comment by HolgerJeromin** [comment]
Just a quick mockup (in Chrome devTools) based on vespucci GUI

![image](https://cloud.githubusercontent.com/assets/2410353/22181804/40af23bc-e094-11e6-8d50-c688d123f032.png)

**Comment by bhousel** [comment]
> Just a quick mockup (in Chrome devTools) based on vespucci GUI

Nice @HolgerJeromin, yes this is what I had in mind too..

**Comment by slhh** [comment]
> How about adding copy and paste action icons right above the All Tags section? 
Do we consider pasting to a single object as sufficient? I think so provided the paste operation is easily accessile and repeatable without copying again. If yes, adding the operations to the all tags sections seems to be a good place.

> This way you will definitely see what you are copying and pasting.
It is a good approch to make the change visible, but there are still some limitations. The change might be in the hidden part of the valuedue to limited edit field width, and detecting all tag changes seems to be difficult in case of many existing tags.

@pnorman We must not assign a shortcut in general, because it would make it a potentially invisible, and therefore unsafe operation. 

We can assign a shortcut to the paste operation in case of an untagged object, the change would be quite visible due to showing the inspector for the feature instead of the feature type selection list. The shortcut  would have a great benefit in this case because the buttons in all tags are not directly accessible due to showing the feature type selection list.

**Comment by slhh** [comment]
@HolgerJeromin Users need to be able to copy tags selectively, but the individual checkboxes waste space which is required for key and value. We can hide the indiviual checkboxes as long as "All" is selected, which should be default. We can also use a toggleable filter icon instead of 'All' and it's checkbox.

There are two different types of usecases for copying tags selectively:

- Selecting tags is required in the copy phase, e.g. when adding multiple new buildings which should all get a basic tag subset.

- Selecting tags is required in the paste phase, e.g. completing address tagging of buildings where overwriting existing tags is not desired.

**Comment by HolgerJeromin** [comment]
Imo no checkbox should be active by default. 
I would not intuitive understand an effect of active  checkboxes while pasting

**Comment by yvecai** [comment]
The user looking for pating tags is looking for productivity,  so checking boxes is not a solution IMO.

**Comment by nighto** [comment]
I think it should just copy paste all tags, boom, done. If user wants not to copy all tags, they could copy paste the object to make a copy, remove the tags thet don't want to copy, and them copy-paste all tags. IMHO it should be as simple as: select object, ctrl-c, select object, ctrl-shift-v.

For tag substitution, I could think of two default behaviours: either simply replace the original clashing property, or join them with `;` and ask the user to correct. I believe the simpler option is to replace.

**Comment by HolgerJeromin** [comment]
> The user looking for pasting tags is looking for productivity,  so checking boxes is not a solution IMO.

I have worked a LOT with Vespucci (1400 changesets) and these checkboxes are very productive :-)

How do you copy or move information from one to another osm object now in iD? Remember (!) all tags and repeatedly switch between both.

**Comment by slhh** [comment]
@nighto 
> If user wants not to copy all tags, they could copy paste the object to make a copy, remove the tags thet don't want to copy, and them copy-paste all tags.

That is a possible workaround, but it generates a dummy object which might get into the database by mistake.
In addition I would worry about users not using the workaround, which can led to copying more tags than really required, and therefore generating an additional risk of overwiting something or adding useless tags.

> IMHO it should be as simple as: select object, ctrl-c, select object, ctrl-shift-v.

This is a simple method to generate garbage. Users would not be aware which tags are copied. There can be tags which are visible in all tags, but not in all fields. All tags might be hidden when pressing Ctrl-C. In addition users would not be aware which tags are overwitten for similar reason.

**Comment by slhh** [comment]
In order to get the required space for the checkboxes without impairing key and value field width, we have the following options:
1. Hide the individual checkboxes while while the global "All" checkbox is active. To make this option useful, we need to make "All" to be active by default, this is the drawback of this option because I would agree with @HolgerJeromin that having all unchecked is the bettter default.
2. Use two rows per tag in the all tags section. This would improve handling of long keys or values anyway.
3. Replace the delete buttons per tag by the checkboxes, add a common delete button on top, make it sensitive to the checkboxes. In this case we don't really need a "Cut" button, because uses can use "Copy" plus " Delete".
4. Keep the current style for all tags by default, at a button at the right side of the "All tags" header line, let the button toggle between the default style and a style according to option 2. This toggle can also unhide the copy-paste menu on top of all tags if desired.
5. Same as above, but toggle between default and style according to option 3.

**Comment by HolgerJeromin** [comment]
@slhh Nice list!
I think 3 would be the best option even if i fear deleting some tags would be much slower...

**Comment by rebio** [comment]
I would like to see this feature. Potlach is able to "repeat tags from previously selected way/node" using R ( http://wiki.openstreetmap.org/wiki/Potlatch_2/Shortcuts ).
This feature in ID would help with nodes in ways where you want to copy the tags to other nodes in ways (eg. crossings). The inexistence of this feature is one major downside for me and a strong argument against using iD at all.

**Comment by jidanni** [comment]
Certainly this should be easy to implement. For me it would just be
```
Ctrl+C	Copy selected features
Ctrl+V	Paste copied features
```
but only copying the tags, not the geometry.
So you might say I am asking for less, not more.

In fact all that needs to be done is pick a vacant control key for this new powered-down version of Crtl+C to be assign to.

**Comment by yvecai** [comment]
You could also paste tag only if a feature is selected.

**Comment by jidanni** [comment]
@yvecai Are you saying that what I want to do is already possible? Is this documented somewhere?

**Comment by yvecai** [comment]
No, just that you wouldn't need another shortcut than ctrl+v.

**Comment by sfkeller** [comment]
+1 for something like "Paste Tags" (Shift+Ctrl+V) in JOSM. I'm leaving it open to implementation what happens when key's clash.

My use case is that I have two text lines (e.g. an address) like this
> addr:street=Main Street
   addr:housenumber=2

Now, I'd like to copy these two tags (Ctrl-C), select an node/object in iD, then paste then as key-values to the node/object.

**Comment by Dulahey** [comment]
I vote for something simple, but ultimately I just want this feature.  However it is implemented, it will speed up production drastically.  It will encourage people to use more thorough tags.

**Comment by ghost07tula** [comment]
I'm really looking forward for this feature. A simple [Ctrl]+[Shift]+[V] for pasting tags would solve a lot of issues.

**Comment by sfkeller** [comment]
I'd like to support Shift+Ctrl+V. And in case of a clash of a tag or a key this key-value gets ignored (a message to the user). 

@bhousel What is left to be sorted out in order to implement this?

**Comment by bhousel** [comment]
@sfkeller I'd like for people to be able to select which tags they are copying.
@HolgerJeromin's mockup above looks great.

**Comment by bhousel** [comment]
Also worth mentioning here - I don't think I have ever wanted this feature.  
But maybe I'm just not clear on how people use it.
I almost never look at the tags in iD.

**Comment by sfkeller** [comment]
I see. But isn't this a bit over-engineered and/or perhaps a second step? When someone selects, he already can decide and has the means e.g. to shift-select ranges, sometimes even being able to pick specific tags.

**Comment by bhousel** [comment]
> When someone selects, he already can decide and has the means e.g. to shift-select ranges, sometimes even being able to pick specific tags.

Shift-click to select ranges doesn't actually work though, right? 

> One of the use cases I know is, that more an more SW - like OSM notes or Maproulette - contain or involve many tags.

I don't understand..

**Comment by sfkeller** [comment]
I realize now that we're speaking of at least two main use cases: 
1. (UC1) Copy tags from another object/source within iD, 
2. (UC2) Copy multiline text (containing hopefully tags) from a text source. 

IMHO both can be handled in one feature - given some good will and backed up with those dozen users who are supporting this issue. 

I try to summarize and specify:
 * GUI general: Add icons (close to Undo/Redo) for Cut | Copy | Paste with hover text showing keyboard short cuts (perhaps stateful changing to disable paste if there's nothing to do).
 * GUI UC1: Show these check boxes for selec/copy as Holger suggested above.
 * Copying UC2: Parse first occurrences of "string=string" in each line of a multiline text to key-value each.
 * Pasting: In case of duplicate tags ignore them, in case of same key add additional values comma-separated. 
* Keyboard short cut: I'm open for "Ctrl-Shift-V" - but don't actually understand yet why there's a Shift modifier key needed.

What do you think?

**Comment by yaugenka** [comment]
On Windows Ctrl + X/C/V are the defaults for cut/copy/paste.

**Comment by slhh** [comment]
@bhousel 
How do you think about https://github.com/openstreetmap/iD/issues/839#issuecomment-275917056 ?

**Comment by omgitsgela** [comment]
I'd like to comment here that this feature would be incredibly helpful for me. I use iD to resolve map notes and fix data quality issues. If someone put business data into a building polygon, but there's actually multiple businesses in that building, it's much better to have those businesses as separate POI within the building boundary. Currently, you have to write it all over from the building polygon tags to the POI.

Another use case is if I am detail mapping, and I want to quickly apply lots of tags to multiple objects. Copying and pasting the actual objects may break geometries or line direction. It would speed things up to be able to just copy the tags over.

Yet another example... a road intersection that needs to be split with a discrete in and out lane, but was originally mapped as a T intersection. Without the tag copy function, you have to write in all the road specifics into the second segment you draw to complete the feature, but copying could do it all automatically, INCLUDING copying over metadata from previous imports.

In short, there's so many reasons to implement a simple copy tags from previous selection feature, I'm surprised we've been stewing on the idea for the last five years without doing anything about it.

**Comment by HolgerJeromin** [comment]
iD is no commercial product where there is a need / possibility to rise the priority of this feature.

No one implemented (a good enough) solution till now.
This is sad, but open source with only volunteers works this way.

**Comment by yaugenka** [comment]
Hooray! This feature was very, very anticipated. Thank you very much @bhousel !

**Comment by jidanni** [comment]
All I know is today I encountered a supermarket marked as a single point.
I wanted to convert it to an area (as it is a giant building on Bing air layer).
I used CTRL+C and then CTRL+V but all I could do is create a 2nd point, not area.
I checked all the Help documentation but nothing was mentioned.
I conclude that currently it is still not possible to copy the tags of a point and put them in an area.

**Comment by nvkelso** [comment]
Draw a polygon, select the point, use the + menu option to combine them?

**Comment by jidanni** [comment]
("jidanni unassigned bhousel 1 hour ago" no jidanni didn't. Maybe GitHub did, but jidanni didn't.)

And all there is when right clicking on https://www.openstreetmap.org/edit?node=4181004666#map=19/24.25173/120.82816 is a trash can, so I don't know what "+" you are talking about.

**Comment by tastrax** [comment]
If you want to combine the two objects then select both (point and the area) and use the C shortcut

https://wiki.openstreetmap.org/wiki/ID/Shortcuts

Not sure what that does with the tags though.

**Comment by tastrax** [comment]
You can also copy the tags from the point object tag TEXT area and paste into the area object tag text area

![image](https://user-images.githubusercontent.com/8775395/61168009-4b5d8200-a58a-11e9-9135-5629c58d693a.png)
